File: android/modules/expo-glasses/android/src/main/java/expo/modules/glasses/ExpoGlassesModule.kt
package expo.modules.glasses

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream

import com.meta.wearable.dat.camera.StreamSession
import com.meta.wearable.dat.camera.startStreamSession
import com.meta.wearable.dat.camera.types.StreamConfiguration
import com.meta.wearable.dat.camera.types.StreamSessionState
import com.meta.wearable.dat.camera.types.VideoFrame
import com.meta.wearable.dat.camera.types.VideoQuality
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.AutoDeviceSelector
import com.meta.wearable.dat.core.types.Permission
import com.meta.wearable.dat.core.types.PermissionStatus
import com.meta.wearable.dat.core.types.RegistrationState

// Android bridge over Meta's Wearables DAT SDK, ported from turbometa's
// WearablesViewModel and matched to the iOS module's JS surface. This compiles
// only once the DAT Maven artifacts resolve, which needs a GitHub token with
// read:packages (see the module README) — so it's built in a dev build, never Expo Go.
class ExpoGlassesModule : Module() {
  private val scope = CoroutineScope(Dispatchers.Main)
  private val selector = AutoDeviceSelector()
  private var session: StreamSession? = null
  private var streamJobs = mutableListOf<Job>()

  private var isAvailable = false
  private var registration = "unavailable"
  private var hasActiveDevice = false
  private var streamState = "stopped"
  private var wantsStreaming = false
  private var lastFrame: Bitmap? = null

  private fun status(): Map<String, Any> = mapOf(
    "isAvailable" to isAvailable,
    "registration" to registration,
    "hasActiveDevice" to hasActiveDevice,
    "streamState" to streamState,
  )

  override fun definition() = ModuleDefinition {
    Name("ExpoGlasses")
    Events("onStatus", "onFrame")

    OnCreate {
      val context = appContext.reactContext ?: return@OnCreate
      try {
        Wearables.initialize(context)
        isAvailable = true
        monitor()
      } catch (e: Throwable) {
        isAvailable = false
        registration = "notRegistered"
      }
      sendEvent("onStatus", status())
    }

    Function("getStatus") { status() }

    AsyncFunction("startRegistration") { promise: Promise ->
      if (registration != "registering" && registration != "registered") {
        appContext.currentActivity?.let { Wearables.startRegistration(it) }
      }
      promise.resolve(null)
    }
    AsyncFunction("startUnregistration") { promise: Promise ->
      appContext.currentActivity?.let { Wearables.startUnregistration(it) }
      promise.resolve(null)
    }
    // The Meta AI OAuth callback is delivered to the Activity's intent on Android,
    // so there's no explicit URL to hand back here; kept for API parity with iOS.
    Function("handleUrl") { _: String -> }

    AsyncFunction("startStreaming") { promise: Promise ->
      wantsStreaming = true
      scope.launch { startStreaming(); promise.resolve(null) }
    }
    AsyncFunction("stopStreaming") { promise: Promise ->
      wantsStreaming = false
      stopStreaming()
      promise.resolve(null)
    }
    AsyncFunction("capturePhoto") { promise: Promise ->
      // Return the latest streamed frame as JPEG. (A dedicated still-capture API
      // exists on the session; the latest frame is the low-risk cross-platform path.)
      val bmp = lastFrame
      if (bmp == null) promise.reject("E_CAPTURE", "No frame available", null)
      else promise.resolve(jpegBase64(bmp))
    }

    OnDestroy { stopStreaming(); scope.coroutineContext[Job]?.cancel() }
  }

  private fun monitor() {
    scope.launch {
      selector.activeDevice(Wearables.devices).collect { device ->
        hasActiveDevice = device != null
        emitStatus()
        if (device != null && wantsStreaming && streamState != "streaming") startStreaming()
      }
    }
    scope.launch {
      Wearables.registrationState.collect { state ->
        registration = when (state) {
          is RegistrationState.Registered -> "registered"
          is RegistrationState.Registering -> "registering"
          else -> "notRegistered"
        }
        emitStatus()
      }
    }
  }

  private suspend fun startStreaming() {
    if (Wearables.checkPermissionStatus(Permission.CAMERA) != PermissionStatus.Granted) {
      // Permission is requested through the Activity's permission launcher; bail
      // until granted rather than crashing.
      return
    }
    stopStreaming()
    val s = Wearables.startStreamSession(selector, StreamConfiguration(VideoQuality.MEDIUM, 24))
    session = s
    streamJobs += scope.launch { s.videoStream.collect { handleFrame(it) } }
    streamJobs += scope.launch {
      s.state.collect { st ->
        streamState = when (st) {
          StreamSessionState.Streaming -> "streaming"
          StreamSessionState.Stopped -> "stopped"
          else -> "waiting"
        }
        emitStatus()
      }
    }
  }

  private fun stopStreaming() {
    streamJobs.forEach { it.cancel() }
    streamJobs.clear()
    session?.stop()
    session = null
  }

  private fun handleFrame(frame: VideoFrame) {
    try {
      val bmp = frame.toBitmap() ?: return
      lastFrame = bmp
      sendEvent("onFrame", mapOf("base64" to jpegBase64(bmp), "width" to bmp.width, "height" to bmp.height))
    } catch (_: Throwable) {}
  }

  // VideoFrame exposes raw bytes; decode defensively. Exact accessors are validated
  // against the SDK during the first dev build.
  private fun VideoFrame.toBitmap(): Bitmap? = try {
    val bytes = this.javaClass.getMethod("getData").invoke(this) as? ByteArray
    bytes?.let { BitmapFactory.decodeByteArray(it, 0, it.size) }
  } catch (_: Throwable) { null }

  private fun jpegBase64(bmp: Bitmap): String {
    val out = ByteArrayOutputStream()
    bmp.compress(Bitmap.CompressFormat.JPEG, 70, out)
    return Base64.encodeToString(out.toByteArray(), Base64.NO_WRAP)
  }

  private fun emitStatus() = sendEvent("onStatus", status())
}

File: android/modules/expo-glasses/ios/ExpoGlassesModule.swift
import ExpoModulesCore
import UIKit

#if canImport(MWDATCore)
import MWDATCore
import MWDATCamera
#endif

// Expo Module bridge over Meta's Wearables DAT SDK. Ports the logic proven in the
// native app's GlassesService (Packages/GlassesKit). All SDK usage is guarded with
// `#if canImport(MWDATCore)` so the module still compiles — and cleanly reports
// `unavailable` — when the SDK isn't linked (e.g. before the SPM package is added,
// or on the simulator where there are no glasses).

public class ExpoGlassesModule: Module {
  private let core = GlassesCore()
  // Snapshot cached off the main actor so the synchronous getStatus() can serve it.
  nonisolated(unsafe) private var lastStatus: [String: Any] = [
    "isAvailable": false, "registration": "unavailable",
    "hasActiveDevice": false, "streamState": "stopped",
  ]

  public func definition() -> ModuleDefinition {
    Name("ExpoGlasses")
    Events("onStatus", "onFrame")

    OnCreate {
      Task { @MainActor in
        self.core.onStatus = { [weak self] status in
          self?.lastStatus = status
          self?.sendEvent("onStatus", status)
        }
        self.core.onFrame = { [weak self] frame in
          self?.sendEvent("onFrame", frame)
        }
        self.core.configure()
      }
    }

    Function("getStatus") { self.lastStatus }

    AsyncFunction("startRegistration") { (promise: Promise) in
      Task { @MainActor in self.core.startRegistration(); promise.resolve(nil) }
    }
    AsyncFunction("startUnregistration") { (promise: Promise) in
      Task { @MainActor in self.core.startUnregistration(); promise.resolve(nil) }
    }
    Function("handleUrl") { (url: String) in
      Task { @MainActor in if let u = URL(string: url) { self.core.handleUrl(u) } }
    }
    AsyncFunction("startStreaming") { (promise: Promise) in
      Task { @MainActor in await self.core.startStreaming(); promise.resolve(nil) }
    }
    AsyncFunction("stopStreaming") { (promise: Promise) in
      Task { @MainActor in await self.core.stopStreaming(); promise.resolve(nil) }
    }
    AsyncFunction("capturePhoto") { (promise: Promise) in
      Task { @MainActor in
        do { promise.resolve(try await self.core.capturePhotoBase64()) }
        catch { promise.reject("E_CAPTURE", error.localizedDescription) }
      }
    }
  }
}

/// Mirrors GlassesService. Kept UI-framework-free so it lives inside the module.
@MainActor
final class GlassesCore {
  // Constructed from the module's nonisolated context; the stored-property
  // defaults are all trivial, so an empty nonisolated init is safe.
  nonisolated init() {}

  var onStatus: (([String: Any]) -> Void)?
  var onFrame: (([String: Any]) -> Void)?

  private(set) var isAvailable = false
  private(set) var registration = "unavailable"
  private(set) var hasActiveDevice = false
  private(set) var streamState = "stopped"
  private(set) var lastError: String?
  private var wantsStreaming = false

  func statusDict() -> [String: Any] {
    var d: [String: Any] = [
      "isAvailable": isAvailable,
      "registration": registration,
      "hasActiveDevice": hasActiveDevice,
      "streamState": streamState,
    ]
    if let lastError { d["lastError"] = lastError }
    return d
  }

  private func emit() { onStatus?(statusDict()) }

  #if canImport(MWDATCore)
  private var wearables: WearablesInterface?
  private var selector: AutoDeviceSelector?
  private var session: StreamSession?
  private var stateToken: AnyListenerToken?
  private var frameToken: AnyListenerToken?
  private var deviceTask: Task<Void, Never>?
  private var registrationTask: Task<Void, Never>?
  private var photoContinuation: CheckedContinuation<Data, Error>?
  private var photoToken: AnyListenerToken?
  private var isProcessingFrame = false
  nonisolated(unsafe) private var foregroundToken: NSObjectProtocol?
  private static var didConfigure = false

  func configure() {
    if !Self.didConfigure {
      do { try Wearables.configure(); Self.didConfigure = true }
      catch { isAvailable = false; registration = "notRegistered"; lastError = "SDK unavailable"; emit(); return }
    }
    let w = Wearables.shared
    let sel = AutoDeviceSelector(wearables: w)
    let cfg = StreamSessionConfig(videoCodec: .raw, resolution: .medium, frameRate: 24)
    let s = StreamSession(streamSessionConfig: cfg, deviceSelector: sel)
    wearables = w; selector = sel; session = s
    isAvailable = true

    registration = map(w.registrationState)
    registrationTask = Task { @MainActor [weak self] in
      guard let self else { return }
      for await state in w.registrationStateStream() { self.registration = self.map(state); self.emit() }
    }
    deviceTask = Task { @MainActor [weak self] in
      guard let self else { return }
      for await device in sel.activeDeviceStream() {
        self.hasActiveDevice = (device != nil); self.emit()
        if device != nil, self.wantsStreaming, self.streamState != "streaming" { await self.startStreaming() }
      }
    }
    stateToken = s.statePublisher.listen { [weak self] state in
      Task { @MainActor [weak self] in self?.updateStreamState(state) }
    }
    frameToken = s.videoFramePublisher.listen { [weak self] frame in
      Task { @MainActor [weak self] in self?.handleFrame(frame) }
    }
    photoToken = s.photoDataPublisher.listen { [weak self] photo in
      Task { @MainActor [weak self] in
        self?.photoContinuation?.resume(returning: photo.data); self?.photoContinuation = nil
      }
    }
    foregroundToken = NotificationCenter.default.addObserver(
      forName: UIApplication.willEnterForegroundNotification, object: nil, queue: .main
    ) { [weak self] _ in
      Task { @MainActor [weak self] in
        guard let self, self.wantsStreaming, self.streamState != "streaming" else { return }
        await self.startStreaming()
      }
    }
    emit()
  }

  func startRegistration() {
    guard let wearables, registration != "registering", registration != "registered" else { return }
    Task { do { try await wearables.startRegistration() } catch { setError(error.localizedDescription) } }
  }

  func startUnregistration() {
    guard let wearables else { return }
    Task { try? await wearables.startUnregistration() }
  }

  func handleUrl(_ url: URL) {
    guard let wearables else { return }
    let comps = URLComponents(url: url, resolvingAgainstBaseURL: false)
    guard comps?.queryItems?.contains(where: { $0.name == "metaWearablesAction" }) == true else { return }
    Task { _ = try? await wearables.handleUrl(url) }
  }

  func startStreaming() async {
    guard let wearables, let session else { return }
    wantsStreaming = true
    do {
      let status = try await wearables.checkPermissionStatus(Permission.camera)
      if status != .granted {
        guard try await wearables.requestPermission(Permission.camera) == .granted else {
          setError("Camera permission denied"); return
        }
      }
      await session.start()
    } catch { setError(error.localizedDescription) }
  }

  func stopStreaming() async {
    wantsStreaming = false
    await session?.stop()
  }

  func capturePhotoBase64(timeout: TimeInterval = 8) async throws -> String {
    guard let session else { throw NSError(domain: "glasses", code: 1) }
    photoContinuation?.resume(throwing: NSError(domain: "glasses", code: 2))
    photoContinuation = nil
    let data: Data = try await withCheckedThrowingContinuation { c in
      self.photoContinuation = c
      session.capturePhoto(format: .jpeg)
      Task { @MainActor [weak self] in
        try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
        if let p = self?.photoContinuation { p.resume(throwing: NSError(domain: "glasses", code: 3)); self?.photoContinuation = nil }
      }
    }
    return data.base64EncodedString()
  }

  private func handleFrame(_ frame: VideoFrame) {
    guard !isProcessingFrame else { return }
    isProcessingFrame = true; defer { isProcessingFrame = false }
    guard let image = frame.makeUIImage(), let jpeg = image.jpegData(compressionQuality: 0.7) else { return }
    onFrame?(["base64": jpeg.base64EncodedString(), "width": image.size.width, "height": image.size.height])
  }

  private func updateStreamState(_ state: StreamSessionState) {
    switch state {
    case .stopped: streamState = "stopped"
    case .streaming: streamState = "streaming"
    case .waitingForDevice, .starting, .stopping, .paused: streamState = "waiting"
    default: streamState = "error"
    }
    emit()
  }

  private func map(_ state: RegistrationState) -> String {
    if state == .registered { return "registered" }
    if state == .registering { return "registering" }
    return "notRegistered"
  }
  #else
  // SDK not linked: everything degrades to the unavailable path.
  func configure() { isAvailable = false; registration = "notRegistered"; emit() }
  func startRegistration() {}
  func startUnregistration() {}
  func handleUrl(_ url: URL) {}
  func startStreaming() async {}
  func stopStreaming() async {}
  func capturePhotoBase64(timeout: TimeInterval = 8) async throws -> String {
    throw NSError(domain: "glasses", code: 0, userInfo: [NSLocalizedDescriptionKey: "Glasses SDK not available"])
  }
  #endif

  private func setError(_ message: String) { lastError = message; emit() }
}
